docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -0,0 +1,312 @@
+    it's still very usable
+* gossip protocol: each lane could select its own gossip protocol.
+  For the future, when we have more than one gossip protocol to choose from.
+* lane/mempool capacity (# of txs, total # of bytes, bytes per tx) **can be different** easily,
+  with little extra complexity on config.
+  * bytes per tx: ok to share one value for MVP... not so sure
+  * \# of txs: probably not all use cases properly addressed if we share a common limit
+  * total # of bytes: if the other two are per-lane, this can just be a "generous" value
+    (and we'd fine tune the other two)
+  * IDEA: Leave this for the end: MVP is viable without it, but use cases less customizable
+* re-check **can be shared** in MVP
+  * with `PrepareProposal`, it becomes _always_ mandatory, so no point in making it per-lane
+
+ADR: This is good info, should probably go at the beginning,
+"These are our decisions on what to implement and what not to.
+The rest of the document only tackles the items reported as "to implement" here.
+
+### Lane coordination
+
+There should be some kind of Lane Coordinator that receives uncategorised txs and assigns them a lane.
+This could be implemented in the reactor or inside `CListMempool`.
+* For an MVP, probably it's easier to put the coordination logic in `CListMempool`.
+* There will be a _native_ or _default_ lane, similar to the _native VLAN_ in an 802.1Q network.
+* `CheckTx` will (optionally) return lane information, and only then it can be put in a lane.
+  (So a tx in a lane has been validated at least once by some node.)
+  * If `CheckTx` provides no lane information: native lane
+* No difference in treatment between TXs coming from RPC, or TXs coming from peers (p2p)
+  * Peers aren't trusted
+  * In any case, the TXs need to be validated first via `CheckTx`
+* The duality lane/priority introduces a powerful indirection
+  * The app can just define the lane of a transaction in `CheckTx`, but the priority of the lane
+    itself can configured (and fine-tuned) elsewhere (also the app?, operators?)
+* Two options for introducing several TX lists:
+  1. Multiple `Clist`s inside `CListMempool`, the cache stays inside `CListMempool`.
+  2. Multiple `CListMempool` instances, the cache needs to be outside `CListMempool` (likely, directly controlled by the reactor)
+* We need to investigate the details of these two options in terms of complexity and risks in the code
+
+ADR: Probably part of [Data flow](#data-flow) section:
+1. some bullets are clearly "Entry flow"
+2. other are general considerations of all flows, or architectural.
+   These are likely to go in a first, introductory subsection of [Data flow](#data-flow) section
+
+### Data flow
+
+* Entry flow:
+  * [Rough description]: the steps when a Tx enters the node are unmodified, except
+    * `CheckTx` (optionally) returns lane information (if TX valid)
+    * The lane information is used to decide to which transaction list the Tx will be added
+* Broadcast flow:
```

Comment:
Using the same routine means, in particular, that a lane with a full send queue would prevent the other lane from being used. Notice that we use a blocking `Send()` on the routine.

Reply:
You're very right. We'll have to add this as a caveat. However, to me, this is a problem to tackle at the p2p level. I would go with what we have, and if our users, when validating the implementation (when it's completed) find the same problems as Evan from Celestia, then we'll have to tackle the problem you mention with top prio.

Reply:
I added a bullet point with this. 🙏 @cason